### ISSUE TITLE
Added support for DER encoding to ECSignature

### ISSUE THREAD

**Issue by stevenroose** [PointyCastle/pointycastle#66]
_From @stevenroose on March 3, 2014 22:7_

Ported this from [dartcoin.ECDSASignature](https://github.com/dartcoin/dartcoin-lib/commit/7164fafe05b9d02f0264bf5c12d3f384d61c8f82).

I implemented the decoding in a quite defensive way by checking validity and throwing a `FormatException` when something invalid is found. If you don't like this, feel free to omit the checks.

I also know you prefer to omit typechecks, so I used `var`'s instead :)

_Copied from original issue: izaera/cipher/pull/66_

**Comment by stevenroose** [comment]
_From @izaera on March 4, 2014 7:27_

Let me think about this one. The way to implement encoding I had in mind was by using auxiliar classes: something like, say, "interface SignatureEncoder" and then a "DERSignatureEncoder" implementation. That way, we decouple abstract objects (signatures, keys, and so on) from their encoding.

In your case I'll probably get your code and move it to such encoder. That way, in the future, when ASN.1 is fully implemented, we can get rid of the ad-hoc implementation and use the valid one.

What do you think?

**Comment by stevenroose** [comment]
_From @izaera on March 4, 2014 7:29_

BTW: the case of RSA was different because there's no "abstract RSA signature", the DER encoding has to be fully embedded in the algorithm because it is then hashed to make the signature, so I cannot decouple it.

**Comment by stevenroose** [comment]
Well, I thought just because it is only temporarily that an inline implementation was good enough.

This is the way I'm doing it in Dartcoin. I'm doing a lot of inline serialization (usually only +-10 lines), because almost all classes have serialization. If I had to create a separate encoding and decoding class for every regular class, I would have almost 3 times as many files.

If you prefer not adding DER support just yet above accepting this PR, I'm perfectly fine with it, I just thought "I did it for my own class, why not add it to cipher's as well".

(BTW, the reason why I have my own ECDSASignature class was because it required DER encoding and also because it has the `ensureCanonical()` method to ensure the signature is in canonical format.
